mod-source-record-manager-server/src/main/java/org/folio/services/flowcontrol/RawRecordsFlowControlServiceImpl.java

```diff
@@ -85,80 +84,99 @@ public void trackChunkReceivedEvent(String tenantId, Integer recordsCount) {
       return;
     }
 
-    if (actualCounterValue < 0) {
-      LOGGER.info("trackRecordCompleteEvent:: Tenant: [{}]. Current value less than zero, because of single record imports or resetting state, back to zero...", tenantId);
-      currentState.put(tenantId, 0);
-      eventProcessedService.resetEventsToProcess(tenantId);
-    } else {
-      currentState.put(tenantId, actualCounterValue);
-    }
-
-    LOGGER.info("--------------- Tenant: [{}]. Current value after complete event: {} ---------------", tenantId, actualCounterValue);
+    decreaseState(tenantId, recordsCount);
+    LOGGER.debug("trackChunkDuplicateEvent:: Tenant: [{}]. Record count: {}, Current state:{}",
+      tenantId, recordsCount, currentState.get(tenantId));
+  }
 
+  private void decreaseState(String tenantId, Integer recordsCount) {
+    initFetchMode(tenantId);
+    decreaseCounterInDb(tenantId, recordsCount);
     resumeIfThresholdAllows(tenantId);
   }
 
   private void resumeIfThresholdAllows(String tenantId) {
-    if (currentState.get(tenantId) <= recordsThreshold) {
-      Collection<KafkaConsumerWrapper<String, String>> rawRecordsReadConsumers = consumersStorage.getConsumersByEvent(DI_RAW_RECORDS_CHUNK_READ.value());
 
-      rawRecordsReadConsumers.forEach(consumer -> {
-        if (consumer.demand() == 0) {
-          consumer.resume();
+    LOGGER.info("resumeIfThresholdAllows :: Tenant: [{}]. Current state:{}, History state: {}",
+      tenantId, currentState.get(tenantId), historyState.get(tenantId));
 
-          LOGGER.info("resumeIfThresholdAllows:: Tenant: [{}]. Kafka consumer - id: {}, subscription - {} is resumed for all tenants, because {} met threshold {}",
-            tenantId, consumer.getId(), DI_RAW_RECORDS_CHUNK_READ.value(), this.currentState, recordsThreshold);
+    historyState.putIfAbsent(tenantId, Pair.of(0,0));
+    if (historyState.get(tenantId).getKey().equals(currentState.get(tenantId))) {
+      Integer previousValue = historyState.get(tenantId).getValue();
+      historyState.put(tenantId, Pair.of(historyState.get(tenantId).getKey(), ++previousValue));
+      LOGGER.debug("resumeIfThresholdAllows :: Tenant: [{}]. Increase counter. Current state:{}, History state: {}",
+        tenantId, currentState.get(tenantId), historyState.get(tenantId));
+    } else {
+      historyState.put(tenantId, Pair.of(currentState.get(tenantId),0));
+    }
+
+    consumersStorage.getConsumersByEvent(DI_RAW_RECORDS_CHUNK_READ.value())
+      .forEach(consumer -> {
+        LOGGER.info("resumeIfThresholdAllows :: Tenant: [{}]. DI_RAW_RECORDS_CHUNK_READ. " +
+            "ConsumerId:{}, Demand:{}, Current state:{}, History state: {}",
+          tenantId, consumer.getId(), consumer.demand(), currentState.get(tenantId), historyState.get(tenantId));
+        if (((consumer.demand() == 0) && (currentState.get(tenantId) < recordsThreshold)) ||
+          (historyState.get(tenantId).getKey() > 0 && historyState.get(tenantId).getValue() > 0)) {
```

Comment:
it think you should get the key and value from historyState at the same time so that your comparison can be consistent.

Reply:
Fixed. Thanks